src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1174 @@
+                allocator,
+                file,
+                file_id,
+                file_offset_start,
+                file_offset_end,
+            );
+    }
+
+    fn readIoUringSubmitAndWait(
+        self: *BufferPool,
+        /// used for temp allocations, and the returned .indices slice
+        allocator: std.mem.Allocator,
+        file: std.fs.File,
+        file_id: FileId,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) !CachedRead {
+        if (builtin.os.tag != .linux)
+            @compileError("io_uring only available on linux - unsupported target");
+
+        const frame_indices = try self.computeFrameIndices(
+            file_id,
+            allocator,
+            file_offset_start,
+            file_offset_end,
+        );
+        errdefer allocator.free(frame_indices);
+
+        // update found frames in the LFU (we don't want to evict these in the next loop)
+        var n_invalid_indices: u32 = 0;
+        for (frame_indices) |f_idx| {
+            if (f_idx == INVALID_FRAME) {
+                n_invalid_indices += 1;
+                continue;
+            }
+            try self.eviction_lfu.insert(self.frames_metadata, f_idx);
+            if (!self.frames_metadata.rc[f_idx].acquire()) {
+                // frame has no handles, but memory is still valid
+                self.frames_metadata.rc[f_idx].reset();
+            }
+        }
+
+        // fill in invalid frames with file data, replacing invalid frames with
+        // fresh ones.
+        var sent_reads: u32 = 0;
+        for (0.., frame_indices) |i, *f_idx| {
+            if (f_idx.* != INVALID_FRAME) continue;
+            // INVALID_FRAME => not found, read fresh and populate
+
+            const frame_aligned_file_offset: FileOffset = @intCast((i * FRAME_SIZE) +
+                (file_offset_start - file_offset_start % FRAME_SIZE));
+            std.debug.assert(frame_aligned_file_offset % FRAME_SIZE == 0);
+
+            f_idx.* = blk: while (true) {
+                if (self.free_list.popOrNull()) |free_idx| {
+                    break :blk free_idx;
+                } else {
+                    try self.evictUnusedFrame();
+                }
+            };
+
+            _ = try self.io_uring.read(
+                f_idx.*,
+                file.handle,
+                .{ .buffer = &self.frames[f_idx.*] },
+                frame_aligned_file_offset,
+            );
+            sent_reads += 1;
+            try self.overwriteDeadFrameInfoNoSize(f_idx.*, file_id, frame_aligned_file_offset);
+            try self.eviction_lfu.insert(self.frames_metadata, f_idx.*);
+        }
+        if (sent_reads != n_invalid_indices) unreachable;
+
+        if (sent_reads > 0) {
+            const n_submitted = try self.io_uring.submit_and_wait(sent_reads);
+            if (n_submitted != sent_reads) unreachable; // did something else submit an event?
```

Comment:
same

Reply:
removed sent_reads in 25ac1b13744cdd660b4643481fc848c7e7dd4532